main/src/library/Region.flix

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2023 Paul Butcher
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Region {
+
+  ///
+  /// Add x to a queue of `Closeable` to be closed when the region exits:
+  ///
+  /// * `Closeable`s are guaranteed to be closed after any threads within the region have been joined.
+  /// * `Closeable`s are not guaranteed to be closed in any particular order.
+  ///
+  pub def closeOnExit(x: a, r: Region[r]): Unit \ {r, IO} with Closeable[a] = 
+    unsafe_cast $ON_EXIT$(() -> Closeable.close(x), r) as _ \ {r, IO}
```

Comment:
I think we need to fix the type inference for ON_EXIT so that this cast disappears.